shared/db/queries/user.js

```diff
@@ -0,0 +1,693 @@
+import type { DBUser, FileUpload, DBThread } from 'shared/types';
+
+export const getUserById = createReadQuery((userId: string) => ({
+  query: db.table('users').get(userId),
+  tags: (user: ?DBUser) => [userId],
+}));
+
+export const getUserByEmail = createReadQuery((email: string) => ({
+  query: db.table('users').getAll(email, { index: 'email' }),
+  process: (users: ?Array<DBUser>) => (users && users[0]) || null,
+  tags: (user: ?DBUser) => (user ? [user.id] : []),
+}));
+
+export const getUserByUsername = createReadQuery((username: string) => ({
+  query: db.table('users').getAll(username, { index: 'username' }),
+  process: (users: ?Array<DBUser>) => (users && users[0]) || null,
+  tags: (user: ?DBUser) => (user ? [user.id] : []),
+}));
+
+export const getUsersByUsername = createReadQuery(
+  (usernames: Array<string>) => ({
+    query: db.table('users').getAll(...usernames, { index: 'username' }),
+    tags: (users: ?Array<DBUser>) => (users ? users.map(({ id }) => id) : []),
+  })
+);
+
+export const getUsers = createReadQuery((userIds: Array<string>) => ({
+  query: db.table('users').getAll(...userIds),
+  tags: (users: ?Array<DBUser>) => (users ? users.map(({ id }) => id) : []),
+}));
+
+export const storeUser = createWriteQuery((user: Object) => ({
+  query: db
+    .table('users')
+    .insert(
+      {
+        ...user,
+        modifiedAt: null,
+      },
+      { returnChanges: true }
+    )
+    .run()
+    .then(() => {
+      identifyQueue.add({ userId: user.id });
+      trackQueue.add({ userId: user.id, event: events.USER_CREATED });
+      sendNewUserWelcomeEmailQueue.add({ user });
+      return Promise.all([user, createNewUsersSettings(user.id)]).then(
+        ([user]) => user
+      );
+    }),
+  invalidateTags: (user: DBUser) => [user.id],
+}));
+
+export const saveUserProvider = createWriteQuery(
+  (
+    userId: string,
+    providerMethod: string,
+    providerId: number,
+    extraFields?: Object
+  ) => ({
+    query: db
+      .table('users')
+      .get(userId)
+      .update(
+        {
+          [providerMethod]: providerId,
+          ...extraFields,
+        },
+        { returnChanges: true }
+      )
+      .run()
+      .then(
+        async ({
+          changes,
+        }: {
+          changes: [{ new_val?: DBUser, old_val?: DBUser }],
+        }) => {
+          const user = changes[0].new_val || changes[0].old_val;
+          if (!user)
+            throw new Error(`Failed to update user with ID ${userId}.`);
+
+          trackQueue.add({
+            userId: userId,
+            event: events.USER_ADDED_PROVIDER,
+            properties: {
+              providerMethod,
+            },
+          });
+
+          identifyQueue.add({ userId });
+
+          return user;
+        }
+      ),
+    invalidateTags: (user: ?DBUser) => (user ? [user.id] : []),
+  })
+);
+
+export const getUserByIndex = createReadQuery(
+  (indexName: string, indexValue: string) => ({
```

Comment:
I realize this is how we have it now in prod, but also realizing that this is a potentially dangerous query if we don't make the `indexName` typesafe to only accept actual indexes we have on this table :P